client_int_test.go

```diff
@@ -155,7 +175,10 @@ func newClient(verbose bool) *boson.Client {
 	builder := buildpacks.NewBuilder()
 	builder.Verbose = verbose
 
-	pusher := docker.NewPusher()
+	pusher, err := docker.NewPusher()
+	if err != nil {
+		panic(err)
```

Comment:
The other constructors call have comment like `TODO: remove error from X constructor`, but is that true? Shouldn't they return error? @lkingland

Reply:
This is just a paradigm which eases the object lifecycle;  By providing constructors with only _data_ necessary for instantiation (sometimes references, if absolutely necessary), and then allowing invocation to be the phase which causes the error.  This can be a nice way to clean up an API. 

It's most certainly not a requirement, and more a note to myself to push things around a bit.  I'll try to whip up a PR to show you what I mean, as it's easier to show rather than explain, and then I can see if you agree with me or not!  Here's the idea, however:

Generally, the object lifecycle is Instantiate, Initialize, Invoke.  By keeping only data (or client references) in the _Instantiate_ and initialize stages (which are combined in the case of constructors), errors can be pushed out to _Invocation_.  This has knock-on effects with the simplification of API client users later in the form of only needing to deal with a single error path.  For a familiar example, think of SQL ORM libraries which tend to not error when you _pass_ them an errant SQL statement, but only when you try to `.Run` them.  The SQL statement is represented as an object which is instantiated, initialized with the raw SQL query, but it's only through invocation that it throws errors.  The lazy evaluation means that there's now only one error path in the code to deal with; the invocation path; which cleans up the code and logic overall.
Again to reiterate; this is just a stylistic choice, more of a note to myself, but perhaps worth the digression, I hope.

Reply:
>For a familiar example, think of SQL ORM libraries which tend to not error when you pass them an errant SQL statement, but only when you try to .Run them.

I think that in case of SQL query they cannot simply and reliably validate the query. What is valid query will depend on SQL dialect on the server and on the DB schema. So I am not sure it's just design choice.

Reply:
>This has knock-on effects with the simplification of API client users later in the form of only needing to deal with a single error path.

Yes it simplify error handling a little bit, still I think it's better to fail fast. Imagine you try to initialize some object and the "constructor" won't return error, then you request some information from user via prompt (like username and pwd).  At the end the call fails and user was bothered unnecessarily. Also other libraries like `docker` or `kn` return errors from "constructors". I am using quotations since I am not sure if there is a thing like constructor in `go`.